source/common/protobuf/message_converter.h

```diff
@@ -0,0 +1,113 @@
+#pragma once
+
+#include "source/common/buffer/buffer_impl.h"
+#include "source/common/buffer/zero_copy_input_stream_impl.h"
+#include "source/common/protobuf/message_converter_utility.h"
+#include "source/common/protobuf/stream_message.h"
+
+#include "grpc_transcoding/message_reader.h"
+#include "proto_field_extraction/message_data/message_data.h"
+
+namespace Envoy::ProtobufMessage {
+
+// MessageConverter is an object that allows conversion from Envoy's gRPC Buffer
+// -> StreamMetadata -> Envoy's gRPC Buffer. API usage example:
+//
+// - Caller has `Envoy::Buffer::Instance` B
+// - Caller calls `AccumulateMessages(B)`, retrieving converted `StreamMessage`
+//   [SM1, SM2, SM3, ...]
+// - Caller is expected to call `ConvertBackToBuffer(SM1)`, then
+//   `ConvertBackToBuffer(SM2)`, etc... to retrieve back Envoy::Buffer
+//   [CB1, CB2, CB3, ...]
+//
+// Note: The caller can run `ConvertBackToBuffer` in any order. They are allowed
+// to convert back buffers out of order.
+//
+// It is invalid to destruct the `MessageConverter` while any data is in the
+// `StreamMessage` form, or while `BytesBuffered` returns a non-zero value.
+// Doing so will result in data loss.
+//
+// MessageConverter is thread-compatible, but it is not tested in multi-threaded
+// environments. This class is meant to be used in Envoy's data plane, where
+// we will have 1 separate MessageConverter instance per worker thread.
+//
+
+class MessageConverter {
+public:
+  MessageConverter(std::unique_ptr<CreateMessageDataFunc> factory)
+      : MessageConverter(std::move(factory), std::numeric_limits<uint32_t>::max()) {}
+
+  explicit MessageConverter(std::unique_ptr<CreateMessageDataFunc> factory, uint32_t buffer_limit);
```

Comment:
It's fine to make this explicit, though we typically don't do that for constructors which take more than one argument.

Reply:
Done.